source/AI.cpp

```diff
@@ -2149,70 +2150,82 @@ void AI::KeepStation(Ship &ship, Command &command, const Body &target)
 
 void AI::Attack(Ship &ship, Command &command, const Ship &target)
 {
-	// First, figure out what your shortest-range weapon is.
-	double shortestRange = 4000.;
-	bool isArmed = false;
-	bool hasAmmo = false;
-	double minSafeDistance = 0.;
-	for(const Hardpoint &hardpoint : ship.Weapons())
-	{
-		const Weapon *weapon = hardpoint.GetOutfit();
-		if(weapon && !hardpoint.IsAntiMissile())
-		{
-			isArmed = true;
-			bool hasThisAmmo = (!weapon->Ammo() || ship.OutfitCount(weapon->Ammo()));
-			hasAmmo |= hasThisAmmo;
-
-			// Exploding weaponry that can damage this ship requires special
-			// consideration (while we have the ammo to use the weapon).
-			if(hasThisAmmo && weapon->BlastRadius() && !weapon->IsSafe())
-				minSafeDistance = max(weapon->BlastRadius() + weapon->TriggerRadius(), minSafeDistance);
-
-			// The missile boat AI should be applied at 1000 pixels range if
-			// all weapons are homing or turrets, and at 2000 if not.
-			double multiplier = (hardpoint.IsHoming() || hardpoint.IsTurret()) ? 1. : .5;
-			shortestRange = min(multiplier * weapon->Range(), shortestRange);
-		}
-	}
-	// If this ship was using the missile boat AI to run away and bombard its
-	// target from a distance, have it stop running once it is out of ammo. This
-	// is not realistic, but it's a whole lot less annoying for the player when
-	// they are trying to hunt down and kill the last missile boat in a fleet.
-	if(isArmed && !hasAmmo)
-		shortestRange = 0.;
-
 	// Deploy any fighters you are carrying.
 	if(!ship.IsYours() && ship.HasBays())
 	{
 		command |= Command::DEPLOY;
 		Deploy(ship, false);
 	}
+	if(ship.GetPersonality().IsRamming())
+	{
+		MoveToAttack(ship, command, target);
+		return;
+	}
 
-	// If this ship has only long-range weapons, or some weapons have a
+	ShipAICache &shipAICache = ship.GetAICache();
+	bool artilleryAI = shipAICache.IsArtilleryAI();
+	double shortestRange = shipAICache.ShortestRange();
+	double shortestArtillery = shipAICache.ShortestArtillery();
+	double minSafeDistance = shipAICache.MinSafeDistance();
+	double totalRadius = ship.Radius() + target.Radius();
+	Point d = target.Position() - ship.Position();
+	double weaponDistance = d.Length() - totalRadius / 3.;
```

Comment:
Why the divide by 3?
Also, `weaponDistance` doesn't seem like an entirely clear name. I'm not even entirely sure what you're using it for right now.

Reply:
I have no idea either

Reply:
Probably an average of the distance between the weapon and the enemy ship